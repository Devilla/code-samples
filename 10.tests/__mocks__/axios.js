let mocked = null;

const noMocks = () => {
  throw('No request mocked');
};

export const mockResponse = (status, statusText, data) => ({
  data,
  status,
  statusText
});

const handleResponse = (mockedUrl, response) =>
  mocked = jest.fn().mockImplementation(({ url }) => {
    mocked = null;

    if (url === mockedUrl) {
      return response;
    }
    throw('Unknown URL: ' + url);
  });

export const clearMock = () => mocked = null;

export const mockRequest = (mockedUrl, status, data) =>
  handleResponse(
    mockedUrl,
    Promise.resolve(mockResponse(status, null, data)));

export const mockRequestError = (mockedUrl, state, error) =>
  handleResponse(
    mockedUrl,
    Promise.reject(mockResponse(state, error, '{}')));

export default {
  request: (params) => (mocked || noMocks)(params),
  get: (url) => (mocked || noMocks)({ url }),
  name: 'mock'
};