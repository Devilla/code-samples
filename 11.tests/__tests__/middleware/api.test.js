import apiMiddleware from '../../app/middleware/api';
import { mockRequestError, mockRequest } from 'axios';
import { API_STARTED, API_FINISHED, API } from '../../app/consts';

const data = { title: 'hello' };

const setData = data => ({
  type: 'SET_DATA',
  payload: data
});

const apiAction = () => ({
  type: API,
  payload: {
    success: setData,
    url: 'fake.json'
  }
});

describe("api middleware", () => {
  let next, dispatch, middleware, dispatchCalls, nextCalls, getState, axiosMock;

  const getStateMock = () => getState ? getState() : {};

  beforeEach(() => {
    getState      = null;
    next          = jest.fn();
    dispatch      = jest.fn();
    dispatchCalls = dispatch.mock.calls;
    nextCalls     = next.mock.calls;
    middleware    = apiMiddleware({ dispatch, getState: getStateMock })(next);
  });

  it('should ignore non API actions', () => {
    const sampleAction = { type: 'SAMPLE_ACTION' };

    middleware(sampleAction);

    expect(dispatchCalls.length).toBe(0);
    expect(nextCalls).toEqual([[sampleAction]]);
  });

  it('should dispatch API_STARTED', () => {
    mockRequest('fake.json', 200, JSON.stringify(data));
    middleware(apiAction());
    expect(dispatchCalls[0]).toEqual([{ type: API_STARTED }]);
  });

  describe('success', () => {
    beforeEach(() => mockRequest('fake.json', 200, JSON.stringify(data)));

    it('should dispatch API_FINISHED', () =>
      middleware(apiAction()).then(() =>
        expect(dispatchCalls[2]).toEqual([{ type: API_FINISHED}])));

    it('should dispatch SET_DATA', () =>
      middleware(apiAction()).then(() =>
        expect(dispatchCalls[1]).toEqual([setData(data)])));
  });

  describe('error', () => {
    beforeEach(() => mockRequestError('fake.json', 404, 'Not found'));

    it('should NOT dispatch API_FINISHED', () =>

      middleware(apiAction()).then(() =>
        expect(dispatchCalls[1][0].type).not.toBe(API_FINISHED)));

    it('should dispatch error', () =>
      middleware(apiAction()).then(() =>
        expect(dispatchCalls[1]).toMatchSnapshot()));
  });

  describe('headers', () => {
    beforeEach(() => axiosMock = mockRequest('fake.json', 200, JSON.stringify(data)));

    it('should set access token if in state', () => {
      getState = () => ({ accessToken: 'xx11' });

      return middleware(apiAction()).then(() => {
        expect(axiosMock.mock.calls[0][0].headers['Access-Token']).toEqual('xx11')
      });
    });

    it('should NOT set access token if NOT in state', () => {
      getState = () => ({ });

      return middleware(apiAction()).then(() => {
        expect(axiosMock.mock.calls[0][0].headers['Access-Token']).toBeUndefined()
      });
    })
  });
});
