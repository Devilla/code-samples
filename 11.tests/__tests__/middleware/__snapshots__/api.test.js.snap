// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api middleware error should dispatch error 1`] = `
Array [
  Object {
    "error": true,
    "payload": [Error: [object Object]],
    "type": "API_ERROR",
  },
]
`;
